### DISCUSSION TITLE
getShopByOwnerUserId returns incorrect data

### DISCUSSION THREAD

**Discussion by milonic** [etsy/open-api#835] [Q&A]
getShopByOwnerUserId is returning the same data over and over, even though different user_id are being correctly passed.

The API is returning data for self, rather than the data for the user_id being passed.

**Comment by Gareth064** [comment]
Has this only just started happening for you?

I used that endpoint recently and it returned the shop of the userid I passed in.

Sent from Outlook for Android<https://aka.ms/AAb9ysg>
________________________________
From: milonic ***@***.***>
Sent: Thursday, January 26, 2023 8:04:28 AM
To: etsy/open-api ***@***.***>
Cc: Subscribed ***@***.***>
Subject: [etsy/open-api] getShopByOwnerUserId returns incorrect data (Discussion #835)


getShopByOwnerUserId is returning the same data over and over, even though different user_id are being correctly passed.

The API is returning data for self, rather than the data for the user_id being passed.

**Comment by milonic** [comment]
No, it's been doing it for quite a while now.

I think it may coincide with when certain user information was suppressed recently, as in this breaking change: https://github.com/etsy/open-api/issues/405#issuecomment-1275082729

It's really useful for us to know if we have sold to a shop (digital items) for copyright infringement checks etc, which are currently rampant on Etsy at the moment.

**Reply by Gareth064** [reply]
Hmm that change was for getting user information not shop information and you are talking about the Get**Shop**ByOwnerId endpoint.

I've just down a test for ownerid 62971666 and it brought back their shop info no problem.

If you have an example ownerId then I will try it too.

**Reply by Gareth064** [reply]
But I think I know what you are trying to do, because I was trying this before.
You were taking the BuyerId from one of your receipts and trying to see if they have a Shop using this endpoint?
If so, that BuyerId is not the same as the OwnerId of a shop.

**Reply by milonic** [reply]
Thanks for getting back to me.

Unfortunately, it doesn't matter what I enter. User_id, shop_id - I also tried  62971666 - it still brings back the same information; my own data.

The reason I think it may be related to the recent change is that a parameter of is_seller used to be available but now it is not and it appears to have been around the same time this change occured.

I really don't know what else I can try, this must be an Etsy issue. Nothing has changed our end.

**Reply by Gareth064** [reply]
If you can get the etsy request ID from the response header then you could email the etsy dev team and ask them to investigate.

**Reply by milonic** [reply]
I've currently got an open ticket with Etsy support, will keep you posted.

**Reply by milonic** [reply]
Finally had a reply from Etsy about this and it's very confusing.

They are telling me that what I am seeing is the correct behavior and these changes were made September 2022. Basically, getShopByOwnerUserId is now a useless end point as it only returns data for "self".

Can you please do me a big favor and test this with shop_id 29069059 and let me know what the shop name is. It should be milonic

What confuses me is that there is literally no documentation on this change anywhere

Many thanks,
Andy

**Reply by Gareth064** [reply]
Yep I will check this later and get back to you.

Etsy response makes no sense because I have an api key associated with my store. And I added my wife's stored to my app and I called that API your talking about, passing the ID of my wife's account and it returned her store. 

If what Etsy is saying is true then it should have returned my store because that's what my API key is associated with

**Reply by Gareth064** [reply]
@milonic I have just realised that '29069059' is a ShopId and not an OwnerId. You need to give the OwnerId as the parameter to this endpoint. You can only get the OwnerId from an Auth Token retrieve after your app has been approved by a shop Owner. 

So for my the access token for my Wifes shop, it looks like
![image](https://user-images.githubusercontent.com/22307223/215868475-79685af4-9af3-4332-8d5e-d41502971729.png)
Which means when I want to get information about her Shop, I use '62971666' as the parameter and you can see in this test app I have, it returns her shop details
![image](https://user-images.githubusercontent.com/22307223/215869136-234cb12a-88e4-4fdd-8ea0-df0ed178cfcc.png)

**Reply by Gareth064** [reply]
When you look up a user in the Web App, you don't get an ID in the address bar like you do when looking at listings or orders etc.
You have to dig deep inside the http requests made between app and server to try and dig out an ID associated to the User, like so.

Using my Wife as an example again.

![image](https://user-images.githubusercontent.com/22307223/215870309-a02962c1-5f4b-40a8-b4e1-fb452f1d1b5c.png)

**Reply by Gareth064** [reply]
Ok scratch everything I said, mostly.

That endpoint returns the shop that the Auth Token belongs to when you make the call to it.

So when you make requests to these Get Endpoints, you have to pass a valid Auth Token, and of course an Auth Token belongs to a Shop/Owner.

So when I make a call to this endpoint using an Auth Token associated to my Wife's Shop, I get her shop details back, no matter which ID I use as a parameter. But then if I make a call to that endpoint using and Auth Token associated with MY shop, then I get my shop returned from it.

So, because that endpoint is auth protected, meaning you have to use a token associated to a shop, it will only return the details of that shop.

You have to pass a numeric value, greater than 0, as the parameter but it doesn't matter what you pass, it will only return the shop associated with the token used.

Here is me passing in a OwnerId of 1, but it still returned my shop (I am using an auth token for my shop)
![image](https://user-images.githubusercontent.com/22307223/215874194-f67da982-05a6-44e4-ae02-661d1ba6a9f2.png)

**Reply by Gareth064** [reply]
I tried calling the endpoint without passing an Auth Token and it failed because it wants the User from the token by the looks of it.

![image](https://user-images.githubusercontent.com/22307223/215875459-8bf4fa18-cbab-4c60-a178-c31e70c48981.png)

The Etsy documentation should be updated to say its protected by OAuth and they should remove the requirement to pass a UserId as a parameter since it doesn't do anything,

**Reply by milonic** [reply]
Hi Gareth,

Thanks so much for your help with this.

I think we can safely say that getShopByOwnerUserId  is now definitely a broken end point and one that is pretty much useless now.

I'm really intrigued as to why Etsy don't want us to match a user to a shop anymore.

Oh well, thanks again.

Andy

**Reply by Gareth064** [reply]
@milonic It's not that you can't match a user to a store at all. You can only match a user to a store when that user has granted you access if that makes sense.

What I think you are looking to do is to take any userId\buyerId and check if they have a store?

**Comment by maxim-skuvault** [comment]
@Gareth064

> [The Etsy documentation] should remove the requirement to pass a UserId as a parameter since it doesn't do anything,

But currently the API does require something to be passed in place of {user_id} in the Url - /v3/application/users/{user_id}/shops. It just doesn't have to be the same user as the one in the passed in OAuth access token. Don't know if it even has to be a valid user anywhere in Etsy, or just a positive integer. In my recent testing both /v3/application/users//shops and /v3/application/users/shops failed with an error "Expected int value for 'user_id'".

**This is a very confusing design.** We've wasted time on this, and I'm sure others have and will.

Not sure what a logical solution is, in light of Etsy choosing an unconventional path of getting the user_id from the access_token, not from the url. Based on the REST API Uri naming best practices, "/v3/application/users/{user_id}/shops" does make sense for this type of endpoint. If they remove {user_id}, then "/v3/application/users/shops" looks like it would return all shops for all users, which it does not. "/v3/application/user/shops" looks weird as well (which user?!?). And I think it should be plural "users", per convention even if one is returned by id. But it might be less confusing to slightly break the convention, than the mess they have now.

Or at the very lease update their docs. For at least 6 months, the doc has been incorrectly saying that this endpoint only requires api_key, not access token ([a different issue](https://github.com/etsy/open-api/discussions/834))
